src/workerd/io/trace.c++

```diff
@@ -711,6 +783,549 @@ tracing::Exception::Exception(rpc::Trace::Exception::Reader reader)
+      info(kj::mv(info)) {}
+
+namespace {
+kj::Maybe<kj::ConstString> readSpanOpenOperationName(rpc::Trace::SpanOpen::Reader& reader) {
+  if (!reader.hasOperationName()) return kj::none;
+  return kj::ConstString(kj::str(reader.getOperationName()));
+}
+
+kj::Maybe<tracing::SpanOpen::Info> readSpanOpenInfo(rpc::Trace::SpanOpen::Reader& reader) {
+  auto info = reader.getInfo();
+  switch (info.which()) {
+    case rpc::Trace::SpanOpen::Info::EMPTY:
+      return kj::none;
+    case rpc::Trace::SpanOpen::Info::FETCH: {
+      return kj::Maybe(tracing::FetchEventInfo(info.getFetch()));
+    }
+    case rpc::Trace::SpanOpen::Info::JSRPC: {
+      return kj::Maybe(tracing::JsRpcEventInfo(info.getJsrpc()));
+    }
+    case rpc::Trace::SpanOpen::Info::CUSTOM: {
+      auto custom = info.getCustom();
+      kj::Vector<tracing::Attribute> attrs(custom.size());
+      for (size_t n = 0; n < custom.size(); n++) {
+        attrs.add(tracing::Attribute(custom[n]));
+      }
+      return kj::Maybe(attrs.releaseAsArray());
+    }
+  }
+  KJ_UNREACHABLE;
+}
+}  // namespace
+
+tracing::SpanOpen::SpanOpen(rpc::Trace::SpanOpen::Reader reader)
+    : operationName(readSpanOpenOperationName(reader)),
+      info(readSpanOpenInfo(reader)) {}
+
+void tracing::SpanOpen::copyTo(rpc::Trace::SpanOpen::Builder builder) {
+  KJ_IF_SOME(name, operationName) {
+    builder.setOperationName(name.asPtr());
+  }
+  KJ_IF_SOME(i, info) {
+    auto infoBuilder = builder.initInfo();
+    KJ_SWITCH_ONEOF(i) {
+      KJ_CASE_ONEOF(fetch, tracing::FetchEventInfo) {
+        fetch.copyTo(infoBuilder.initFetch());
+      }
+      KJ_CASE_ONEOF(jsrpc, tracing::JsRpcEventInfo) {
+        jsrpc.copyTo(infoBuilder.initJsrpc());
+      }
+      KJ_CASE_ONEOF(custom, tracing::CustomInfo) {
+        auto customBuilder = infoBuilder.initCustom(custom.size());
+        for (size_t n = 0; n < custom.size(); n++) {
+          custom[n].copyTo(customBuilder[n]);
+        }
+      }
+    }
+  }
+}
+
+tracing::SpanOpen tracing::SpanOpen::clone() {
+  static constexpr auto cloneInfo =
+      [](kj::Maybe<Info>& info) -> kj::Maybe<tracing::SpanOpen::Info> {
+    return info.map([](Info& info) -> tracing::SpanOpen::Info {
+      KJ_SWITCH_ONEOF(info) {
+        KJ_CASE_ONEOF(fetch, tracing::FetchEventInfo) {
+          return fetch.clone();
+        }
+        KJ_CASE_ONEOF(jsrpc, tracing::JsRpcEventInfo) {
+          return jsrpc.clone();
+        }
+        KJ_CASE_ONEOF(custom, tracing::CustomInfo) {
+          kj::Vector<tracing::Attribute> attrs(custom.size());
+          for (size_t n = 0; n < custom.size(); n++) {
+            attrs.add(custom[n].clone());
+          }
+          return attrs.releaseAsArray();
+        }
+      }
+      KJ_UNREACHABLE;
+    });
+  };
+  return SpanOpen(
+      operationName.map([](auto& str) { return kj::ConstString(kj::str(str)); }), cloneInfo(info));
+}
+
+tracing::SpanClose::SpanClose(EventOutcome outcome): outcome(outcome) {}
```

Comment:
I guess, this can be moved to the header file.

Reply:
It could but I prefer to keep it with the rest of the impl. Just a style difference